lib/_http_parser.js

```diff
@@ -0,0 +1,895 @@
+var FLAG_SHOULD_KEEP_ALIVE = 1 << 7;
+var FLAG_CONNECT_METHOD = 1 << 8;
+var FLAG_ANY_UPGRADE = FLAG_UPGRADE | FLAG_CONNECTION_UPGRADE;
+
+function HTTPParser(type) {
+  if (type === RESPONSE) {
+    this.type = type;
+    this._state = STATE_STATUS_LINE;
+  }
+  this.headers = [];
+}
+HTTPParser.prototype.type = REQUEST;
+HTTPParser.prototype._state = STATE_REQ_LINE;
+HTTPParser.prototype._err = null;
+HTTPParser.prototype._flags = 0;
+HTTPParser.prototype._contentLen = null;
+HTTPParser.prototype._nbytes = null;
+HTTPParser.prototype._nhdrbytes = 0;
+HTTPParser.prototype._nhdrpairs = 0;
+HTTPParser.prototype._buf = '';
+HTTPParser.prototype.httpMajor = 1;
+HTTPParser.prototype.httpMinor = null;
+HTTPParser.prototype.maxHeaderPairs = 2000;
+HTTPParser.prototype.method = null;
+HTTPParser.prototype.url = null;
+HTTPParser.prototype.statusCode = null;
+HTTPParser.prototype.statusText = null;
+
+HTTPParser.prototype.onHeaders = null;
+HTTPParser.prototype.onBody = null;
+HTTPParser.prototype.onComplete = null;
+
+HTTPParser.prototype.onIncoming = null;
+HTTPParser.prototype.incoming = null;
+HTTPParser.prototype.socket = null;
+
+HTTPParser.prototype.reinitialize = function(type) {
+  this.execute = this._executeStartLine;
+  this.type = type;
+  if (type === REQUEST)
+    this._state = STATE_REQ_LINE;
+  else
+    this._state = STATE_STATUS_LINE;
+
+  this._err = null;
+  this._flags = 0;
+  this._contentLen = null;
+  this._nbytes = null;
+  this._nhdrbytes = 0;
+  this._nhdrpairs = 0;
+  this._buf = '';
+
+  // Common properties
+  this.headers = [];
+  this.httpMajor = 1;
+  this.httpMinor = null;
+  this.maxHeaderPairs = 2000;
+
+  // Request properties
+  this.method = null;
+  this.url = null;
+
+  // Response properties
+  this.statusCode = null;
+  this.statusText = null;
+};
+HTTPParser.prototype.finish = function() {
+  var state = this._state;
+  if (state === STATE_BODY_EOF) {
+    this.execute = this._executeBodyIgnore;
+    this._state = STATE_COMPLETE;
+    this.onComplete && this.onComplete();
+  } else if (this.execute !== this._executeError &&
+             state !== STATE_REQ_LINE &&
+             state !== STATE_STATUS_LINE &&
+             state !== STATE_COMPLETE) {
+    return this._setError('Invalid EOF state');
+  }
+};
+HTTPParser.prototype.close = function() {};
+HTTPParser.prototype.pause = function() {};
+HTTPParser.prototype.resume = function() {};
+HTTPParser.prototype._setError = function(msg) {
+  var err = new Error(msg);
+  this.execute = this._executeError;
+  this._err = err;
+  return err;
+};
+HTTPParser.prototype._headersEnd = function() {
+  var flags = this._flags;
+  var type = this.type;
+  var method = this.method;
+  var upgrade = ((flags & FLAG_ANY_UPGRADE) === FLAG_ANY_UPGRADE ||
+                 (flags & FLAG_CONNECT_METHOD) > 0);
+  var contentLen = this._contentLen;
+  var httpMajor = this.httpMajor;
+  var httpMinor = this.httpMinor;
+  var statusCode = this.statusCode;
```

Comment:
Usually local shortcuts do not help much and even can make things slower. I would not use it, if not benhmarked every case. Also note, that `contentLen` var now can be `null` or `Number`. It's better to fix such things before benchmarking to be sure in correct result.

PS. this note apply to all similar places.